### PR TITLE
Add rubocop-i18n and config

### PR DESCRIPTION
**Pull request by kbarrette**
For discussion:

This is how I currently have `rubocop-i18n` [configured in `catalog-management-service`](https://github.com/salsify/catalog-management-service/pull/1364) and it's working nicely for me.

I've got a [PR open](https://github.com/puppetlabs/rubocop-i18n/pull/57) on the gem to incorporate the monkey patch I have here, but I'm not sure if they're accepting contributions.

Note that I only require the gem in the config yml, so this is off by default, users can opt in by changing their rubocop config to something like:

```yaml
inherit_gem:
  salsify_rubocop:
    - conf/rubocop_rails50.yml
    - conf/rubocop_i18n.yml
```

### CODE REVIEW

File: lib/rubocop/cop/salsify/i18n.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module I18n
+      module RailsI18n
+        class DecorateString < Cop
```
Comment (kbarrette) [inline lib/rubocop/cop/salsify/i18n.rb]: Not quite sure which filename to use for this class

File: lib/rubocop/cop/salsify/i18n.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+          # Original: https://github.com/puppetlabs/rubocop-i18n/blob/v3.0.0/lib/rubocop/cop/i18n/rails_i18n/decorate_string.rb#L138
+          # This patch adds the parent_is_logger? check
+          def check_for_parent_decorator(node)
```
Comment (will89) [inline lib/rubocop/cop/salsify/i18n.rb]: I think you can simplify this approach and make it easier for future maintainability by using `Module.prepend`.
Something like

```
Module MyPatch
  def check_for_parent_decorator(node)
    return if parent_is_logger?(node.parent)

    super(node)
  end

  private

  def parent_is_logger?(parent)
    parent.respond_to?(:method_name) && [:debug, :info, :warn, :error].include?(parent.method_name)
  end
end

RuboCop::Cop::I18n::RailsI18n::DecorateString.prepend(MyPatch)
```

Comment (kbarrette) [reply lib/rubocop/cop/salsify/i18n.rb]: Ah yes, I really should move past ruby 1.9 😄 

Will fix!

**Review by will89 — APPROVED** [review]

**Review by skarger — APPROVED** [review]
I tried using this branch in brand-space and amazingly it didn't flag problems. I'll be interested to see what examples arise.

**Comment by erikkessler1** [comment]
> I tried using this branch in brand-space and amazingly it didn't flag problems. I'll be interested to see what examples arise.

Did you remember to add the `- conf/rubocop_i18n.yml` to your configuration? I also thought somehow content-flow-service had no violations which I found hard to believe, but adding that configuration raised 900+ violations 😅 

If we're not going to make it enabled by default, should we document that you need to add that configuration somewhere or even include it in our rails template so new services at least know the option exists?

**Comment by kbarrette** [comment]
> > I tried using this branch in brand-space and amazingly it didn't flag problems. I'll be interested to see what examples arise.
> 
> Did you remember to add the `- conf/rubocop_i18n.yml` to your configuration? I also thought somehow content-flow-service had no violations which I found hard to believe, but adding that configuration raised 900+ violations 😅
> 
> If we're not going to make it enabled by default, should we document that you need to add that configuration somewhere or even include it in our rails template so new services at least know the option exists?


Yeah, make sure to add this to your rubocop config:

```yaml
inherit_gem:
  salsify_rubocop:
    - conf/rubocop_rails50.yml
    - conf/rubocop_i18n.yml
```

Documenting this sounds like a good idea - will do 😄

**Comment by kbarrette** [comment]
@skarger @erikkessler1 added some docs

**Review by erikkessler1 — APPROVED** [review]

**Comment by skarger** [comment]
Good point @erikkessler1, on my first look I read over that part too fast. Turns out brand-space has 288 violations with those lines added.

@kbarrette this gets at my original goal in testing it: There are quite a few flagged strings that can't or shouldn't be internationalized. Non-exhaustive examples:

* GraphQL field documentation ([example](https://github.com/salsify/brand-space/blob/master/app/graphql/concerns/locale_mapping_fields.rb))
* Log messages that aren't passed directly inline to `Rails.logger` ([example](https://github.com/salsify/brand-space/blob/fe9681ec92eb1738d025f7da4cea640116b3f7f2/app/service_objects/amazon_us/aplus_postgres.rb#L96-L99))
* Clockwork schedule config lines like ['Monday 07:00'](https://github.com/salsify/brand-space/blob/fe9681ec92eb1738d025f7da4cea640116b3f7f2/lib/clock.rb#L6)
* Names in emails published for `email-service`, like [recipients: [{ email_address: 'EPMproductdemos@bestbuy.com', name: 'EPM Product Demos' }]](https://github.com/salsify/brand-space/blob/fe9681ec92eb1738d025f7da4cea640116b3f7f2/app/jobs/brand_space/best_buy_email_job.rb#L73)
* SQL strings ([example](https://github.com/salsify/brand-space/blob/fe9681ec92eb1738d025f7da4cea640116b3f7f2/app/service_objects/reporting.rb#L66-L69))

The fact that it's opt-in helps, and then of course we could add ignore-lines for all these, but we'd end up with a lot.

So I wonder if rubocop is the right fit for this need. Maybe, if we identify the types of strings that should be internationalized, we could externalize just those?

I realize that therefore wouldn't catch things people forgot to include, like say there's a new result message for a background job that someone types inline in English, there would be no linting to detect that. Maybe we could mitigate that by extracting some libraries that built in i18n for the areas that need it, e.g. for how to return GraphQL error messages?

All that said, I suppose it's fine if a team wants to use rubocop for this, so opt-in doesn't do much harm.

**Comment by kbarrette** [comment]
@skarger I was hoping this would be useful to a wider selection of teams - if the brand-space experience is going to be the common situation, you're right, this is probably not useful. Inside catalog-management-service I've just got a couple of rubocop exceptions where we're sending slack messages, making API calls, etc, so this worked really well for us.

I guess the question is: are other teams going to use this? If not, let's not muddy up this gem with it.

**Comment by skarger** [comment]
@kbarrette hard to say what % of teams would want it.

One possibility is to wait on this PR for a short time until the i18n efforts yield some more info / best practices about how to migrate a backend service, and then judge whether to merge this.

The opt in part allows for teams to experiment, which could be helpful to determine the utility.

**Pull request closed without merging**